src/libcore/slice/mod.rs

```diff
@@ -2364,14 +2328,66 @@ impl<'a, T> IntoIterator for &'a mut [T] {
     }
 }
 
-#[inline]
-fn size_from_ptr<T>(_: *const T) -> usize {
-    mem::size_of::<T>()
-}
-
 // The shared definition of the `Iter` and `IterMut` iterators
 macro_rules! iterator {
-    (struct $name:ident -> $ptr:ty, $elem:ty, $mkref:ident) => {
+    (struct $name:ident -> $ptr:ty, $elem:ty, $raw_mut:tt, $( $mut_:tt )*) => {
+        impl<'a, T> $name<'a, T> {
+            // Helper function for creating a slice from the iterator.
+            #[inline(always)]
+            fn make_slice(&self) -> &'a [T] {
+                unsafe { from_raw_parts(self.ptr, self.len()) }
+            }
+
+            // Helper function for moving the start of the iterator forwards by `offset` elements,
+            // returning the old start.
+            // Unsafe because the offset must be in-bounds or one-past-the-end.
+            #[inline(always)]
+            unsafe fn post_inc_start(&mut self, offset: isize) -> * $raw_mut T {
+                if mem::size_of::<T>() == 0 {
+                    self.end = (self.end as isize).wrapping_sub(offset) as * $raw_mut T;
```

Comment:
`wrapping_sub` looks like a typo in a method named `inc`?

Reply:
I thought so too at first, but no, for ZST slices the start pointer should not change as the iterator advances, and this wrapping_sub is decrementing *the length*. However, since it's apparently somewhat subtle, there should be a comment spelling this out.

Reply:
Yeah in fact I used `wrapping_add` at first and wondered why everything exploded. ;)

I'll add a comment.